engine/src/main/java/org/pentaho/di/trans/steps/webservices/WebService.java

```diff
@@ -0,0 +1,1122 @@
+
+    try {
+
+      StringBuilder sb = new StringBuilder( Math.max( 16, is.available() ) );
+      char[] tmp = new char[ 4096 ];
+
+      try {
+        InputStreamReader reader = new InputStreamReader( is, encoding != null ? encoding : "UTF-8" );
+        for ( int cnt; ( cnt = reader.read( tmp ) ) > 0; ) {
+          sb.append( tmp, 0, cnt );
+        }
+      } finally {
+        is.close();
+      }
+      return sb.toString();
+
+    } catch ( Exception e ) {
+      throw new KettleStepException( "Unable to read web service response data from input stream", e );
+    }
+
+  }
+
+  private void processRows( InputStream anXml, Object[] rowData, RowMetaInterface rowMeta,
+                            boolean ignoreNamespacePrefix, String encoding ) throws KettleException {
+    // Just to make sure the old transformations keep working...
+    //
+    if ( meta.isCompatible() ) {
+      compatibleProcessRows( anXml, rowData, rowMeta, ignoreNamespacePrefix, encoding );
+      return;
+    }
+
+    // First we should get the complete string
+    // The problem is that the string can contain XML or any other format such as HTML saying the service is no longer
+    // available.
+    // We're talking about a WEB service here.
+    // As such, to keep the original parsing scheme, we first read the content.
+    // Then we create an input stream from the content again.
+    // It's elaborate, but that way we can report on the failure more correctly.
+    //
+    String response = readStringFromInputStream( anXml, encoding );
+
+    try {
+
+      // What is the expected response object for the operation?
+      //
+      DocumentBuilderFactory documentBuilderFactory = XMLParserFactoryProducer.createSecureDocBuilderFactory();
+      documentBuilderFactory.setNamespaceAware( true );
+
+      DocumentBuilder documentBuilder = documentBuilderFactory.newDocumentBuilder();
+
+      Document doc = documentBuilder.parse( new InputSource( new StringReader( response ) ) );
+
+      Node envelopeNode = doc.getFirstChild();
+      String nsPrefix = envelopeNode.getPrefix();
+      Node bodyNode = XMLHandler.getSubNode( envelopeNode, nsPrefix + ":Body" );
+      if ( bodyNode == null ) {
+        XMLHandler.getSubNode( envelopeNode, nsPrefix + ":body" ); // retry, just in case!
+      }
+
+      // Create a few objects to help do the layout of XML snippets we find along the way
+      //
+      Transformer transformer = null;
+      /*
+       * as of BACKLOG-4068, explicit xalan factory references have been deprecated; we use the javax.xml factory
+       * and let java's SPI determine the proper transformer implementation. In addition, tests has been made to
+       * safeguard that https://github.com/pentaho/pentaho-kettle/commit/3b57f7a9aac657fe77cc4f08e8d4287fcccbc073
+       * continues working as intended
+       */
+
+      TransformerFactory transformerFactory = XMLParserFactoryProducer.createSecureTransformerFactory();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Using TransformerFactory without appropriate configuration can lead to XXE